core/src/main/scala/org/http4s/blaze/pipeline/stages/SSLStage.scala

```diff
@@ -31,202 +38,203 @@ final class SSLStage(engine: SSLEngine, maxWrite: Int = 1024 * 1024)
-    * after such an operation is stashed.
-    */
-  private def sslHandshake(data: ByteBuffer, r: HandshakeStatus): Unit = {
-    val start = System.nanoTime
-    handshakeQueue.synchronized {
-      @tailrec
-      def sslHandshakeLoop(data: ByteBuffer, r: HandshakeStatus): Unit =
-        r match {
-          case HandshakeStatus.NEED_UNWRAP =>
-            val o = getScratchBuffer(maxBuffer)
-            val r = engine.unwrap(data, o)
-
-            r.getStatus match {
-              case Status.OK =>
-                sslHandshakeLoop(data, r.getHandshakeStatus)
-
-              case Status.BUFFER_UNDERFLOW =>
-                channelRead().onComplete {
-                  // use `sslHandshake` to reacquire the lock
-                  case Success(b) =>
-                    sslHandshake(concatBuffers(data, b), r.getHandshakeStatus)
-                  case Failure(t) => handshakeFailure(t)
-                }(trampoline)
-
-              case Status.CLOSED => // happens if the handshake fails for some reason
-                handshakeFailure(new SSLException(s"SSL Closed"))
-
-              case Status.BUFFER_OVERFLOW =>
-                handshakeFailure(util.bug(s"Unexpected status: ${Status.BUFFER_OVERFLOW}"))
-            }
-
-          case HandshakeStatus.NEED_TASK =>
-            runTasks()
-            sslHandshakeLoop(data, engine.getHandshakeStatus)
-
-          case HandshakeStatus.NEED_WRAP =>
-            val o = getScratchBuffer(maxBuffer)
-            val r = engine.wrap(emptyBuffer, o)
-            o.flip()
-
-            if (r.bytesProduced() < 1)
-              logger.warn(s"SSL Handshake WRAP produced 0 bytes.\n$r")
-
-            channelWrite(copyBuffer(o)).onComplete {
-              // use `sslHandshake` to reacquire the lock
-              case Success(_) => sslHandshake(data, r.getHandshakeStatus)
-              case Failure(t) => handshakeFailure(t)
-            }(trampoline)
-
-          // Finished with the handshake: continue what we were doing.
-          case _ =>
-            assert(readLeftover == null)
-            readLeftover = data
-            val pendingOps = handshakeQueue.result(); handshakeQueue.clear()
-            logger.trace(s"Submitting backed up ops: $pendingOps")
-            pendingOps.foreach {
-              case DelayedRead(sz, p) => doRead(sz, p)
-              case DelayedWrite(d, p) => continueWrite(d, p)
-            }
-        }
-
-      try sslHandshakeLoop(data, r)
-      catch {
-        case t: SSLException =>
-          logger.warn(t)("SSLException in SSL handshake")
-          handshakeFailure(t)
-
-        case t: Throwable =>
-          logger.error(t)("Error in SSL handshake")
-          handshakeFailure(t)
+  // Perform the SSL Handshake and then continue any pending operations.
+  // There should be at least one pending operation as this is only started
+  // after such an operation is stashed.
+  private[this] def sslHandshake(data: ByteBuffer, r: HandshakeStatus): Unit = {
```

Comment:
What's the reason for `private` => `private[this]`?

Reply:
It's a more appropriate notion of private. Vanill `private` allows the SSLStage instances touch each others methods while `private[this]` restricts access to only `this` instance.

I think in practice 99.9% of usages of `private` in classes/traits should really be `private[this]`.

Reply:
ah interesting. I actually didn't know that `private[this]` restricted it to the instance. I've never run into it since I've never tried calling another `private` from another one of the same object.

Thanks!

Reply:
`private[this]` also is direc access to a field instead of through a method, which can save a few picoseconds here and there.